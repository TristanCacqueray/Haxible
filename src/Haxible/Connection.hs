-- | This module contains the logic to interact with the python wrapper
module Haxible.Connection (Connections (..), TaskCall (..), withConnections) where

import Control.Exception (bracket)
import Data.Aeson (Value (Object), eitherDecodeStrict, encode)
import Data.Aeson.Key qualified
import Data.Aeson.KeyMap qualified
import Data.Bifunctor (first)
import Data.ByteString (hGetLine, toStrict)
import Data.ByteString.Char8 (hPutStrLn)
import Data.Hashable (Hashable)
import Data.Pool qualified
import Data.Text (Text)
import Data.Text qualified as Text
import Data.Typeable (Typeable)
import GHC.Generics (Generic)
import GHC.IO.Handle (hFlush)
import Say
import System.IO (Handle, hClose)
import System.Process.Typed

data TaskCall = TaskCall
  { -- | The playbook attributes, such as `hosts` or `become`
    playAttrs :: [(Text, Value)],
    -- | The task object, e.g `{"file": {"path": "/etc/zuul"}}`
    taskObject :: Value,
    -- | A list of variables, such as `register` names or `include_vars`
    env :: [(Text, Value)]
  }
  deriving (Eq, Show, Typeable, Generic, Hashable)

-- | A connection run converts a TaskCall into a (result code, result value)
newtype Connections = Connections {run :: TaskCall -> IO (Int, Value)}

-- | Creates the Python interpreters.
withConnections :: Int -> FilePath -> (Connections -> IO ()) -> IO ()
withConnections count inventory callback =
  bracket (Data.Pool.newPool poolConfig) Data.Pool.destroyAllResources go
  where
    go pool = do
      putStrLn "Pool ready"

      let runTask :: TaskCall -> Process Handle Handle () -> IO (Int, Value)
          runTask taskCall p = do
            let callParams = [mkObj taskCall.playAttrs, taskCall.taskObject, mkObj taskCall.env]
            say $ " â–¶ Calling " <> Text.pack (show taskCall)
            hPutStrLn (getStdin p) (toStrict $ encode callParams)
            hFlush (getStdin p)
            output <- hGetLine (getStdout p)
            case eitherDecodeStrict output of
              Right res -> pure res
              Left err -> error $ show output <> ": " <> err

      callback (Connections $ Data.Pool.withResource pool . runTask)

    mkObj = Object . Data.Aeson.KeyMap.fromList . map (first Data.Aeson.Key.fromText)

    poolConfig :: Data.Pool.PoolConfig (Process Handle Handle ())
    poolConfig =
      Data.Pool.PoolConfig
        { createResource = do
            startProcess
              . setStdin createPipe
              . setStdout createPipe
              $ proc "python" ["./app/wrapper.py", inventory],
          freeResource = \p -> do
            hClose (getStdin p)
            stopProcess p,
          poolCacheTTL = 3600,
          poolMaxResources = count
        }
