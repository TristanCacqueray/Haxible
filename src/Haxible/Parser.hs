module Haxible.Parser (decodePlaybook, renderScript) where

import Data.Aeson
import Data.Aeson.Key qualified
import Data.Aeson.KeyMap qualified as HM (lookup, toList)
import Data.ByteString (ByteString, toStrict)
import Data.Foldable
import Data.Text (Text)
import Data.Text qualified as Text
import Data.Text.Encoding
import Data.Vector (toList)
import Data.Yaml

newtype Playbook = Playbook [Task] deriving (Eq, Show)

data Task = Task
  { name :: Maybe Text,
    action :: Text,
    attributes :: Value,
    requires :: [Text],
    register :: Maybe Text,
    loop :: Maybe [Text]
  }
  deriving (Eq, Show)

decodePlaybook :: ByteString -> Playbook
decodePlaybook bs = case Data.Yaml.decodeEither' bs of
  Right (Array v) -> Playbook (reverse $ fst $ foldl' addRequirement ([], []) $ decodeTask <$> Data.Vector.toList v)
  x -> error $ "Invalid pb: " <> show x

addRequirement :: ([Task], [Text]) -> Task -> ([Task], [Text])
addRequirement (xs, available) task = (task {requires = requires} : xs, newAvailable)
  where
    requires = findRequirements task.attributes
    findRequirements :: Value -> [Text]
    findRequirements v = case v of
      String x -> case filter (`Text.isInfixOf` x) available of
        [] -> []
        requirement -> requirement
      Object x -> concatMap findRequirements x
      Array x -> concatMap findRequirements x
      _ -> []
    newAvailable = case task.register of
      Just reg -> reg : available
      Nothing -> available

decodeTask :: Value -> Task
decodeTask (Object obj) =
  let register = decodeTextValue "register"
      name = decodeTextValue "name"
      loop = case HM.lookup "loop" obj of
        Just (Array x) -> traverse decodeText (Data.Vector.toList x)
        _ -> Nothing
      (action, attributes) = case filter (\(n, _) -> taskAttribute n) (HM.toList obj) of
        [(n, v)] -> (Data.Aeson.Key.toText n, v)
        [] -> error "Missing task"
        xs -> error $ "Unknown task: " <> show xs
      requires = []
   in Task {name, action, attributes, requires, register, loop}
  where
    taskAttribute n = Data.Aeson.Key.toText n `notElem` ["name", "register", "loop"]
    decodeTextValue k = decodeText =<< HM.lookup k obj
    decodeText v = case v of
      String x -> Just x
      _ -> Nothing
decodeTask v = error $ "Unexpected task shape: " <> show v

renderScript :: Playbook -> Text
renderScript (Playbook tasks) =
  Text.unlines $
    [ "{-# LANGUAGE QuasiQuotes, ApplicativeDo, OverloadedStrings #-}",
      "{- cabal:",
      "build-depends: base, haxible",
      "ghc-options: -threaded -rtsopts -with-rtsopts=-N -with-rtsopts=-T",
      "-}",
      "module Main (main) where\n",
      "import Haxible.Eval\n",
      "main :: IO ()",
      "main = runHaxible playbook\n",
      "playbook :: AnsibleHaxl ()",
      "playbook = do"
    ]
      <> (mappend "  " . renderCode <$> tasks)
      <> ["  pure ()"]

quote :: Text -> Text
quote = Text.cons '"' . flip Text.snoc '"'

renderCode :: Task -> Text
renderCode task = Text.unwords (registerBind <> taskCall)
  where
    registerBind = case task.register of
      Just v -> [v, "<-"]
      Nothing -> []

    taskCall = case task.loop of
      Nothing -> directCall
      Just xs -> ["traverse", "(\\item -> "] <> directCall <> [") " <> Text.pack (show xs)]

    directCall = ["runTask", "(" <> Text.pack (show task.name) <> ")", quote task.action, arguments]

    arguments = case requirements of
      [] -> attributes
      xs -> "(renderTemplates [" <> Text.intercalate ", " (mkArg <$> xs) <> "] " <> attributes <> ")"

    attributes = "[json| " <> decodeUtf8 (Data.ByteString.toStrict $ Data.Aeson.encode task.attributes) <> " |]"

    requirements = case task.loop of
      Nothing -> task.requires
      Just _ -> "item" : task.requires

    mkArg v = "(" <> quote v <> ", " <> v <> ")"
