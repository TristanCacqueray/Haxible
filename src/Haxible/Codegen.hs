module Haxible.Codegen (renderScript) where

import Data.Aeson (encode)
import Data.Aeson.Key qualified
import Data.Aeson.KeyMap qualified
import Data.Text qualified as Text
import Haxible.Normalize
import Haxible.Prelude

renderScript :: FilePath -> [Definition] -> Text
renderScript inventory defs =
  Text.unlines $
    [ "-- Generated with haxible",
      "{-# LANGUAGE QuasiQuotes, ApplicativeDo, OverloadedStrings #-}",
      "{- cabal:",
      "build-depends: base, haxible",
      "ghc-options: -threaded -rtsopts -with-rtsopts=-N -with-rtsopts=-T",
      "-}",
      "module Main (main) where\n",
      "import Haxible.Eval\n",
      "main :: IO ()",
      "main = runHaxible " <> quote (from inventory) <> " (playbook [] [])\n"
    ]
      <> concatMap renderDefinition defs

renderDefinition :: Definition -> [Text]
renderDefinition def =
  [ def.name <> " :: Vars -> Vars -> AnsibleHaxl [Value]",
    def.name <> " playAttrs baseEnv = do"
  ]
    <> (mappend "  " <$> map renderExpr def.exprs)
    <> ["  pure $ " <> outputList, ""]
  where
    outputList = Text.intercalate " <> " (toOutput <$> def.exprs)
    toOutput expr = case expr.term of
      -- Module call produces a single value
      ModuleCall _ -> "[" <> from expr.binder <> "]"
      -- Otherwise we got a list of values
      _ -> from expr.binder

renderExpr :: Expr -> Text
renderExpr e = from e.binder <> " <- " <> Text.unwords finalExpr
  where
    requirements = textList $ (\req -> "(" <> quote req.name <> ", " <> mkOrigin req.origin <> ")") <$> e.requirements
    mkOrigin = \case
      Direct n -> from n
      Nested n i -> from n <> " !! " <> from (show i)
      LoopVar -> "__haxible_loop_item"

    finalExpr = case e.loop of
      Just (Array xs) -> mkTraverse $ textList (embedJSON <$> toList xs)
      Just (String v) -> mkTraverse $ "(envLoop " <> quote v <> " " <> paren (requirements <> " <> baseEnv") <> ")"
      Just _ -> error $ "Invalid loop expression: " <> show e.loop
      Nothing -> callExpr
      where
        traverser = case e.term of
          ModuleCall _ -> "traverseLoop"
          DefinitionCall _ -> "traverseInclude"
        mkTraverse arg = [traverser, "(\\__haxible_loop_item -> "] <> callExpr <> [") ", arg]

    callExpr = case e.term of
      ModuleCall CallModule {module_, params, taskAttrs} ->
        [ "runTask",
          "playAttrs",
          quote module_,
          embedJSON (obj $ [(module_, params)] <> taskAttrs),
          paren (requirements <> " <> baseEnv")
        ]
      DefinitionCall CallDefinition {name, playAttrs, baseEnv} ->
        [ name,
          paren (textList (mkJsonArg <$> playAttrs) <> " <> playAttrs"),
          paren (requirements <> " <> " <> textList (mkJsonArg <$> baseEnv) <> " <> baseEnv")
        ]

paren :: Text -> Text
paren = Text.cons '(' . flip Text.snoc ')'

quote :: Text -> Text
quote = Text.cons '"' . flip Text.snoc '"'

embedJSON :: Value -> Text
embedJSON v = "[json|" <> unsafeFrom (Data.Aeson.encode v) <> "|]"

mkJsonArg :: (Text, Value) -> Text
mkJsonArg (n, v) = "(" <> quote n <> ", " <> embedJSON v <> ")"

textList :: [Text] -> Text
textList xs = "[" <> Text.intercalate ", " xs <> "]"

obj :: [(Text, Value)] -> Value
obj = Object . Data.Aeson.KeyMap.fromList . map (first Data.Aeson.Key.fromText)
