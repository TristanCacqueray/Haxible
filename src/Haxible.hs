-- | The Haxible library entrypoint.
module Haxible where

import Haxible.Codegen
import Haxible.Import
import Haxible.Normalize
import Haxible.Prelude
import Haxible.Syntax
import System.Process.Typed (ExitCode (..), proc, readProcess, runProcess_)

-- | Transform an inventory and playbook into a haskell script.
compile :: FilePath -> FilePath -> IO Text
compile inventory playPath = do
  basePlays <- Haxible.Syntax.decodeFile playPath
  plays <- traverse (Haxible.Import.resolveImport playPath) basePlays
  let exprs = Haxible.Normalize.normalizePlaybook plays
  pure $ renderScript inventory playPath exprs []

-- | Execute the haskell script with cabal.
execute :: FilePath -> IO ()
execute script = runProcess_ (proc "cabal" ["run", "-O0", script])

validate :: FilePath -> IO ()
validate script = do
  (res, stdout, stderr) <- readProcess (proc "runhaskell" [script])
  case res of
    ExitSuccess -> pure ()
    ExitFailure x -> error $ "Process failed with " <> show x <> "\n" <> unsafeFrom stdout <> "\n" <> unsafeFrom stderr
