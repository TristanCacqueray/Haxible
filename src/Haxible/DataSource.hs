-- | This module contains the Haxl logic
module Haxible.DataSource (AnsibleHaxl, initHaxibleState, dataFetch, TaskReq (..)) where

import Control.Concurrent (QSem, newQSem, signalQSem, waitQSem)
import Control.Concurrent.Async (async)
import Control.Exception (SomeException, bracket_, try)
import Control.Monad (void)
import Data.Aeson (Value)
import Data.Hashable (Hashable (hashWithSalt))
import Data.List qualified as List
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import Data.Text qualified as Text
import Data.Typeable (Typeable)
import Haxible.Connection (Python (..))
import Haxl.Core
import Say

type AnsibleHaxl a = GenHaxl () () a

data TaskReq a where
  RunTask :: Maybe Text -> Text -> Value -> TaskReq Value
  deriving (Typeable)

deriving instance Eq (TaskReq a)

deriving instance Show (TaskReq a)

instance ShowP TaskReq where showp = show

instance Hashable (TaskReq a) where
  hashWithSalt s (RunTask _ n v) = hashWithSalt s (0 :: Int, n, v)

instance StateKey TaskReq where
  data State TaskReq = AnsibleState
    { python :: Python,
      semaphore :: QSem
    }

instance DataSourceName TaskReq where
  dataSourceName _ = "Ansible"

instance DataSource u TaskReq where
  fetch = fetchTask

initHaxibleState :: Int -> Python -> IO (State TaskReq)
initHaxibleState threads python = do
  semaphore <- newQSem threads
  pure $ AnsibleState {python, semaphore}

fetchTask :: State TaskReq -> Flags -> u -> PerformFetch TaskReq
fetchTask state _flags _user =
  BackgroundFetch $ \reqs -> do
    say $ "[+] Batching " <> Text.pack (show (List.length reqs)) <> " tasks"
    mapM_ (fetchAsync state.python state.semaphore) reqs

fetchAsync :: Python -> QSem -> BlockedFetch TaskReq -> IO ()
fetchAsync python sem (BlockedFetch req rvar) =
  void $
    async $
      bracket_ (waitQSem sem) (signalQSem sem) $ do
        e <- Control.Exception.try $ runTaskReq python req
        case e of
          Left ex -> putFailure rvar (ex :: SomeException)
          Right a -> putSuccess rvar a

runTaskReq :: Python -> TaskReq a -> IO a
runTaskReq python (RunTask name action attr) = do
  say $ " â–¶ Calling " <> fromMaybe "unnamed task" name
  python.call action attr
