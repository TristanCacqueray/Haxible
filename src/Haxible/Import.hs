-- | This module contains the logic to resolve roles and include tasks
module Haxible.Import
  ( resolveImport,
    BasePlay (..),
    BaseTask (..),
    Play,
    Task,
    TaskValue (..),
    RoleValue (..),
  )
where

import Haxible.Prelude
import Haxible.Syntax

type Importer a = ReaderT Env IO a

data Env = Env
  { source :: FilePath,
    history :: [FilePath]
  }
  deriving (Eq, Show)

type Play = BasePlay Task

type Task = BaseTask TaskValue

data TaskValue
  = Module Value
  | Role RoleValue
  | Tasks Text [Task]
  | Facts Vars
  | CacheableFacts Value Vars
  | Block BlockValue
  deriving (Eq, Show)

data RoleValue = RoleValue
  { tasks :: [Task],
    defaults :: [(Text, Value)],
    name :: Text
  }
  deriving (Eq, Show)

data BlockValue = BlockValue
  { tasks :: [Task],
    rescues :: [Task]
  }
  deriving (Eq, Show)

resolveTask :: TaskSyntax -> Importer Task
resolveTask task = do
  taskValue <- case task.module_ of
    "include_role" -> includeRole
    "include_tasks" -> includeTasks
    "set_fact" -> setFact
    "block" -> block
    _ -> pure $ Module task.params
  pure $ task {params = taskValue}
  where
    withFile fp go = do
      hist <- asks history
      when (fp `elem` hist) $
        error $ "Cyclic import detected: " <> show fp <> " already in " <> show hist
      r <- decodeFile fp
      local (const $ Env fp (fp : hist)) $ go r

    getRolePath name path = do
      source <- asks source
      pure $ takeDirectory source </> "roles" </> name </> path
    includeRole = do
      let role_name = from $ fromMaybe "missing name" $ preview (key "name" . _String) $ task.params
          name = from role_name
      role_path <- getRolePath role_name $ "tasks" </> "main.yaml"
      role_defaults <- getRolePath role_name $ "defaults" </> "main.yaml"
      JsonVars defaults <- decodeFile role_defaults
      withFile role_path $ \baseTasks -> do
        tasks <- traverse resolveTask baseTasks
        pure $ Role RoleValue {name, tasks, defaults}

    includeTasks = do
      source <- asks source
      let task_name = from $ fromMaybe "missing name" $ preview _String $ task.params
          task_path = takeDirectory source </> task_name
      withFile task_path $ \baseTasks -> do
        Tasks (from task_name) <$> traverse resolveTask baseTasks

    block = do
      tasks <- resolveBlock task.params
      rescues <- resolveBlock (fromMaybe Null (lookup "rescue" task.attrs))
      pure $ Block BlockValue {tasks, rescues}
    resolveBlock :: Value -> Importer [Task]
    resolveBlock = traverse resolveTask . unwrapJSON . fromJSON
    unwrapJSON = \case
      Error e -> error $ "Unexpected json: " <> e
      Success a -> a

    setFact = do
      let JsonVars vars = unwrapJSON . fromJSON $ task.params
      pure $ case lookup "cacheable" vars of
        Just v -> CacheableFacts v (filter (\var -> fst var /= "cacheable") vars)
        Nothing -> Facts vars

-- | Transform a 'PlaySyntax' into a resolved 'Play'
resolveImport :: FilePath -> PlaySyntax -> IO Play
resolveImport source (BasePlay baseTasks attrs) = do
  tasks <- runReaderT (traverse resolveTask baseTasks) (Env source [])
  pure $ BasePlay {tasks, attrs}
