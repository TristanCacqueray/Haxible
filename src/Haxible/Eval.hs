-- | This module contains the evaluation logic
module Haxible.Eval
  ( AnsibleHaxl,
    runHaxible,
    runTask,
    json,
    envLoop,
    traverseLoop,
    traverseInclude,
    cleanVar,
    Value,
    Vars,
  )
where

import Data.Aeson hiding (json)
import Data.Aeson.KeyMap qualified
import Data.Aeson.QQ
import Data.Default (def)
import Data.Functor.Identity (runIdentity)
import Data.Vector qualified
import Haxible.Connection
import Haxible.DataSource
import Haxible.Prelude
import Haxl.Core hiding (env)
import Language.Haskell.TH.Quote qualified
import Text.Ginger

type Vars = [(Text, Value)]

json :: Language.Haskell.TH.Quote.QuasiQuoter
json = aesonQQ

envLoop :: Text -> [(Text, Value)] -> [Value]
envLoop n env = runGinger context template
  where
    lookupVar var = maybe def toGVal (lookup var env)
    gvarToList g = case g.asList of
      Just xs -> case traverse fromGValEither xs of
        Right x -> x
        Left err -> error $ "Couldn't resolve loop list: " <> err
      Nothing -> error $ "Invalid loop variable: " <> show g

    context = makeContext' lookupVar gvarToList Nothing
    template :: Template SourcePos
    template = case runIdentity (parseGinger (const undefined) Nothing (from n)) of
      Right tmpl -> tmpl
      Left e -> error $ "Template fail: " <> show e

traverseLoop :: Applicative f => (a -> f Value) -> [a] -> f Value
traverseLoop f xs = loopResult <$> traverse f xs

traverseInclude :: Applicative f => (a -> f [Value]) -> [a] -> f [Value]
traverseInclude f xs = concat <$> traverse f xs

loopResult :: [Value] -> Value
loopResult xs = Object $ Data.Aeson.KeyMap.fromList attrs
  where
    attrs =
      [ ("results", Array (Data.Vector.fromList $ cleanVar <$> xs)),
        ("msg", "All items completed"),
        ("skipped", Bool $ all (fromMaybe False . preview (key "skipped" . _Bool)) xs),
        ("changed", Bool $ any (fromMaybe False . preview (key "changed" . _Bool)) xs)
      ]

runTask :: [(Text, Value)] -> Text -> Value -> [(Text, Value)] -> AnsibleHaxl Value
runTask playAttrs module_ taskObject baseEnv = dataFetch (RunTask (TaskCall {playAttrs, taskObject, env, module_}))
  where
    env = concatMap checkManyHost baseEnv
    -- When a task run on many host, we register a single variable with all the results,
    -- thus when accessing the variable, we need to lookup the current host result.
    checkManyHost (k, v)
      | preview (key "__haxible_many_hosts" . _Bool) v == Just True =
          let many_k = k <> "__haxible"
           in [ (k, String $ "{{ " <> many_k <> "[ansible_host] }}"),
                (many_k, cleanVar v)
              ]
      | otherwise = [(k, cleanVar v)]

runHaxible :: FilePath -> AnsibleHaxl [Value] -> IO ()
runHaxible inventory action = withConnections 5 inventory $ \connections -> do
  ansibleState <- initHaxibleState connections
  ansibleEnv <- initEnv (stateSet ansibleState stateEmpty) ()
  xs <- runHaxl ansibleEnv action
  putStrLn "\nReport:"
  traverse_ printResult xs
  putStrLn "Done."
  where
    printResult :: Value -> IO ()
    printResult v = putStrLn $ "[ok]: " <> unsafeFrom (encode v)
