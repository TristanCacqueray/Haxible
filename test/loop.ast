[ Definition
    { name = "playbook"
    , requires = []
    , provides = []
    , outputs = Environment
        { getEnv = [] }
    , exprs =
        [ Expr
            { binder = Binder "resultsLocalhost0"
            , requires = []
            , provides =
                [ Resource
                    { name = Binder "debugCallTaskInALoop0"
                    , dep = Register "loop_res"
                    }
                ]
            , outputs = Right []
            , requirements = []
            , loop = Nothing
            , term = DefinitionCall
                ( CallDefinition
                    { name = "playLocalhost0"
                    , playAttrs =
                        [
                            ( "hosts"
                            , String "localhost"
                            )
                        ]
                    , baseEnv = []
                    }
                )
            }
        ]
    }
, Definition
    { name = "playLocalhost0"
    , requires = []
    , provides =
        [ Resource
            { name = Binder "debugCallTaskInALoop0"
            , dep = Register "loop_res"
            }
        ]
    , outputs = Environment
        { getEnv =
            [
                ( Binder "debugCallTaskInALoop0"
                , Right
                    [ Resource
                        { name = Binder "debugCallTaskInALoop0"
                        , dep = Register "loop_res"
                        }
                    ]
                )
            ,
                ( Binder "debug0"
                , Right []
                )
            ]
        }
    , exprs =
        [ Expr
            { binder = Binder "debugCallTaskInALoop0"
            , requires = []
            , provides =
                [ Resource
                    { name = Binder "debugCallTaskInALoop0"
                    , dep = Register "loop_res"
                    }
                ]
            , outputs = Right
                [ Resource
                    { name = Binder "debugCallTaskInALoop0"
                    , dep = Register "loop_res"
                    }
                ]
            , requirements =
                [ Requirement
                    { name = "item"
                    , origin = LoopVar
                    }
                ]
            , loop = Just
                ( Array
                    [ String "A"
                    , String "B"
                    , String "C"
                    ]
                )
            , term = ModuleCall
                ( CallModule
                    { module_ = "debug"
                    , params = Object
                        ( fromList
                            [
                                ( "msg"
                                , String "loop {{ item }}"
                                )
                            ]
                        )
                    , taskAttrs =
                        [
                            ( "name"
                            , String "Call task in a loop"
                            )
                        ]
                    }
                )
            }
        , Expr
            { binder = Binder "debug0"
            , requires =
                [ Resource
                    { name = Binder "debugCallTaskInALoop0"
                    , dep = Register "loop_res"
                    }
                ]
            , provides = []
            , outputs = Right []
            , requirements =
                [ Requirement
                    { name = "loop_res"
                    , origin = Direct
                        ( Binder "debugCallTaskInALoop0" )
                    }
                ]
            , loop = Nothing
            , term = ModuleCall
                ( CallModule
                    { module_ = "debug"
                    , params = Object
                        ( fromList
                            [
                                ( "msg"
                                , String "loop result is {{ loop_res }}"
                                )
                            ]
                        )
                    , taskAttrs = []
                    }
                )
            }
        ]
    }
]