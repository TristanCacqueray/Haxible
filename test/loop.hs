#!/usr/bin/env cabal
-- Generated with haxible
{-# LANGUAGE QuasiQuotes, ApplicativeDo, OverloadedStrings #-}
{- cabal:
build-depends: base, haxible
ghc-options: -threaded -rtsopts -with-rtsopts=-N -with-rtsopts=-T
-}
module Main (main) where

import Haxible.Eval

main :: IO ()
main = runHaxible "inventory.yaml" (playbook [] [])

playbook :: Vars -> Vars -> AnsibleHaxl [Value]
playbook playAttrs baseEnv = do
  resultsLocalhost0 <- playLocalhost0 ([("hosts", [json|"localhost"|])] <> playAttrs) ([] <> [] <> baseEnv)
  pure $ resultsLocalhost0

playLocalhost0 :: Vars -> Vars -> AnsibleHaxl [Value]
playLocalhost0 playAttrs baseEnv = do
  debugCallTaskInALoop0 <- traverseLoop (\__haxible_loop_item ->  runTask playAttrs "debug" [json|{"debug":{"msg":"loop {{ item }}"},"name":"Call task in a loop"}|] ([("item", __haxible_loop_item)] <> baseEnv) )  [[json|"A"|], [json|"B"|], [json|"C"|]]
  debug0 <- runTask playAttrs "debug" [json|{"debug":{"msg":"loop result is {{ loop_res }}"}}|] ([("loop_res", debugCallTaskInALoop0)] <> baseEnv)
  pure $ [debugCallTaskInALoop0] <> [debug0]

