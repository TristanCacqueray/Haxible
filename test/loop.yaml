- hosts: localhost
  tasks:
    - name: Call task in a loop
      debug:
        msg: "loop {{ item }}"
      loop: ["A", "B", "C"]
      register: loop_res

    - debug:
        msg: "loop result is {{ loop_res }}"
