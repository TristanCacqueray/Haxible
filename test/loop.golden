(
    [ Definition
        { name = "playbook"
        , requires = []
        , provides = []
        , outputs = Environment
            { getEnv = [] }
        , exprs =
            [ Expr
                { binder = Binder "resultsLocalhost0"
                , requires = []
                , provides =
                    [ Resource
                        { name = Binder "debugCallTaskInALoop0"
                        , dep = Register "loop_res"
                        }
                    ]
                , outputs = Right []
                , requirements = []
                , loop = Nothing
                , term = DefinitionCall
                    ( CallDefinition
                        { name = "playLocalhost0"
                        , playAttrs =
                            [
                                ( "hosts"
                                , String "localhost"
                                )
                            ]
                        , baseEnv = []
                        }
                    )
                }
            ]
        }
    , Definition
        { name = "playLocalhost0"
        , requires = []
        , provides =
            [ Resource
                { name = Binder "debugCallTaskInALoop0"
                , dep = Register "loop_res"
                }
            ]
        , outputs = Environment
            { getEnv =
                [
                    ( Binder "debugCallTaskInALoop0"
                    , Right
                        [ Resource
                            { name = Binder "debugCallTaskInALoop0"
                            , dep = Register "loop_res"
                            }
                        ]
                    )
                ,
                    ( Binder "debug0"
                    , Right []
                    )
                ]
            }
        , exprs =
            [ Expr
                { binder = Binder "debugCallTaskInALoop0"
                , requires = []
                , provides =
                    [ Resource
                        { name = Binder "debugCallTaskInALoop0"
                        , dep = Register "loop_res"
                        }
                    ]
                , outputs = Right
                    [ Resource
                        { name = Binder "debugCallTaskInALoop0"
                        , dep = Register "loop_res"
                        }
                    ]
                , requirements =
                    [ Requirement
                        { name = "item"
                        , origin = LoopVar
                        }
                    ]
                , loop = Just
                    ( Array
                        [ String "A"
                        , String "B"
                        , String "C"
                        ]
                    )
                , term = ModuleCall
                    ( CallModule
                        { module_ = "debug"
                        , params = Object
                            ( fromList
                                [
                                    ( "msg"
                                    , String "loop {{ item }}"
                                    )
                                ]
                            )
                        , taskAttrs =
                            [
                                ( "name"
                                , String "Call task in a loop"
                                )
                            ]
                        }
                    )
                }
            , Expr
                { binder = Binder "debug0"
                , requires =
                    [ Resource
                        { name = Binder "debugCallTaskInALoop0"
                        , dep = Register "loop_res"
                        }
                    ]
                , provides = []
                , outputs = Right []
                , requirements =
                    [ Requirement
                        { name = "loop_res"
                        , origin = Direct
                            ( Binder "debugCallTaskInALoop0" )
                        }
                    ]
                , loop = Nothing
                , term = ModuleCall
                    ( CallModule
                        { module_ = "debug"
                        , params = Object
                            ( fromList
                                [
                                    ( "msg"
                                    , String "loop result is {{ loop_res }}"
                                    )
                                ]
                            )
                        , taskAttrs = []
                        }
                    )
                }
            ]
        }
    ]
, "-- Generated with haxible
  {-# LANGUAGE QuasiQuotes, ApplicativeDo, OverloadedStrings #-}
  {- cabal:
  build-depends: base, haxible
  ghc-options: -threaded -rtsopts -with-rtsopts=-N -with-rtsopts=-T
  -}
  module Main (main) where

  import Haxible.Eval

  main :: IO ()
  main = runHaxible "inventory.yaml" (playbook [] [])

  playbook :: Vars -> Vars -> AnsibleHaxl [Value]
  playbook playAttrs baseEnv = do
    resultsLocalhost0 <- playLocalhost0 ([("hosts", [json|"localhost"|])] <> playAttrs) ([] <> [] <> baseEnv)
    pure $ resultsLocalhost0

  playLocalhost0 :: Vars -> Vars -> AnsibleHaxl [Value]
  playLocalhost0 playAttrs baseEnv = do
    debugCallTaskInALoop0 <- traverseLoop (\__haxible_loop_item ->  runTask playAttrs "debug" [json|{"debug":{"msg":"loop {{ item }}"},"name":"Call task in a loop"}|] ([("item", __haxible_loop_item)] <> baseEnv) )  [[json|"A"|], [json|"B"|], [json|"C"|]]
    debug0 <- runTask playAttrs "debug" [json|{"debug":{"msg":"loop result is {{ loop_res }}"}}|] ([("loop_res", debugCallTaskInALoop0)] <> baseEnv)
    pure $ [debugCallTaskInALoop0] <> [debug0]

  "
)